### PR TITLE
cleanup tokens

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 1]
```diff
@@ -118,8 +112,7 @@ const AssetSelectorItem = ({wallet, tokenInfo, onSelect, matcher}: AssetSelector
 
         <View style={{flex: 1, padding: 4}}>
           <Text numberOfLines={1} ellipsizeMode="middle" style={{color: COLORS.BLUE_LIGHTER}} testID="tokenInfoText">
-            {/* eslint-disable-next-line @typescript-eslint/strict-boolean-expressions */}
-            {tokenInfo.ticker || tokenInfo.name || '-'}
+            {tokenInfo.ticker ?? tokenInfo.name ?? '-'}
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx:115]: better types

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 2]
```diff
@@ -60,7 +54,7 @@ export const AssetSelectorScreen = () => {
       </View>
 
       <FlatList
-        data={sortedTokenInfos}
+        data={sortTokenInfos({wallet, tokenInfos})}
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx:57]: shared sorting

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 3]
```diff
@@ -47,7 +41,7 @@ export const AssetSelectorScreen = () => {
   return (
     <SafeAreaView edges={['left', 'right', 'bottom']} style={{flex: 1, backgroundColor: 'white'}}>
       <View style={{paddingTop: 16, paddingHorizontal: 16}}>
-        <SearchInput onChangeText={(text) => onChangeMatcher(normalize(text))} />
+        <SearchInput onChangeText={(text) => onChangeMatcher(text)} />
```
Comment (wolverineks) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx:44]: moved into `matches`

File: src/TxHistory/AssetList/AssetList.tsx [hunk 4]
```diff
@@ -55,7 +49,7 @@ export const AssetList = (props: Props) => {
 
       <FlatList
         {...props}
-        data={sortedTokenInfos}
+        data={sortTokenInfos({wallet, tokenInfos})}
```
Comment (wolverineks) [inline src/TxHistory/AssetList/AssetList.tsx:52]: shared sorting

File: src/TxHistory/AssetList/AssetList.tsx [hunk 5]
```diff
@@ -87,7 +81,7 @@ const AssetItem = ({tokenInfo, onPress}: AssetItemProps) => {
 
       <Middle>
         <Text numberOfLines={1} ellipsizeMode="middle" style={styles.tokenInfo} testID="tokenInfoText">
-          {tokenInfo.ticker ?? tokenInfo.name}
+          {tokenInfo.ticker ?? tokenInfo.name ?? '-'}
```
Comment (wolverineks) [inline src/TxHistory/AssetList/AssetList.tsx:84]: better types

File: src/yoroi-wallets/cardano/api/api.ts [hunk 6]
```diff
@@ -113,42 +110,6 @@ export const getTokenInfo = async (tokenId: string, apiUrl: string) => {
   return entry ? tokenInfo(entry) : fallbackTokenInfo(tokenId)
 }
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/api/api.ts:112]: moved to utils

File: src/yoroi-wallets/cardano/api/api.ts [hunk 7]
```diff
@@ -166,58 +127,13 @@ export const fetchCurrentPrice = async (currency: CurrencySymbol, config: Backen
   return response.ticker.prices[currency]
 }
 
-export const hexToAscii = (hex: string) => Buffer.from(hex, 'hex').toLocaleString()
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/api/api.ts:169]: moved to utils

File: src/yoroi-wallets/cardano/api/api.ts [hunk 8]
```diff
@@ -233,13 +149,25 @@ type Signature = {
...
+  value: T | undefined
 }
 
 const parseTokenRegistryEntry = (data: unknown) => {
   return isTokenRegistryEntry(data) ? data : undefined
 }
 
 const isTokenRegistryEntry = (data: unknown): data is TokenRegistryEntry => {
-  return !!data && typeof data === 'object' && 'name' in data && 'assetName' in data && 'policyId' in data
+  const candidate = data as TokenRegistryEntry
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/api/api.ts:160]: fixed type check

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx [hunk 9]
```diff
@@ -152,16 +146,18 @@ const Icon = (props) => (
...
 const SearchInput = (props) => {
   const strings = useStrings()
 
   return <TextInput {...props} label={strings.searchLabel} />
 }
 
 const matches = (tokenInfo: TokenInfo, matcher: string) =>
-  Object.values(tokenInfo).some((key) => normalize(String(key)).includes(normalize(matcher)))
+  Object.values(tokenInfo)
+    .filter((value): value is string | number => value != null)
```
Comment (michalszorad) [inline src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx:157]: let's use strict checks and also would recommend to use `typeof value === 'string' || typeof value === 'number'` to make the TypeGuard more future proof if we add a new property to `TokenInfo` that has a different type from `string` `number` and `undefined`

File: src/yoroi-wallets/cardano/api/utils.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  const tokenSubject = toTokenSubject(tokenIdentifier)
+  const assetName = hexToAscii(tokenSubject.slice(56)) || undefined
+  return assetName
+}
+
+export const toTokenSubject = (tokenIdentifier: string) => tokenIdentifier.replace('.', '')
+export const toTokenId = (tokenIdentifier: string) => {
+  const tokenSubject = toTokenSubject(tokenIdentifier)
+  return `${tokenSubject.slice(0, 56)}.${tokenSubject.slice(56)}`
+}
```
Comment (michalszorad) [inline src/yoroi-wallets/cardano/api/utils.ts:64]: Is this just an abstraction so we can move to use AssetId as a token id? Also, would it be worth re-using the existing functions? Eg

```
export const toTokenId = (tokenIdentifier: string) => {
  const tokenSubject = toTokenSubject(tokenIdentifier)
  return `${toPolicyId(tokenIdentifier)}.${toAssetNameHex(tokenIdentifier)}`
}
```

**Review by michalszorad — APPROVED** [review]

File: src/utils/sorting.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import {YoroiWallet} from '../yoroi-wallets'
+import {TokenInfo} from '../yoroi-wallets/types'
+
+export const sortTokenInfos = ({wallet, tokenInfos}: {wallet: YoroiWallet; tokenInfos: Array<TokenInfo>}) =>
```
Comment (stackchain) [inline src/utils/sorting.ts:4]: @wolverineks in my head is something that we should export within `yoroi-wallets`, so we sort on all platforms in the same way. Maybe moving the `/utils` to `yoroi-wallets/` ?

File: src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx (hunk 9 above)
Comment (stackchain) [reply src/Send/AssetSelectorScreen/AssetSelectorScreen.tsx:157]: not sure about it, `NaN` will resolve to "NaN" anyways. maybe a clear parser?

File: src/TxHistory/AssetList/AssetList.tsx (hunk 5 above)
Comment (stackchain) [reply src/TxHistory/AssetList/AssetList.tsx:84]: mental note: come with something to unify this logic, maybe a utility function `displayName`

File: src/yoroi-wallets/cardano/api/utils.ts (hunk 10 above)
Comment (stackchain) [reply src/yoroi-wallets/cardano/api/utils.ts:64]: `toPolicyId` makes sense, but I think the `tokenIdentifier` name is already in hex.

**Review by stackchain — COMMENTED** [review]
⬆️

File: src/yoroi-wallets/cardano/api/utils.ts (hunk 10 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/api/utils.ts:64]: i was going for...
```
tokenIdentifier: tokenId | tokenSubject
tokenId:         policyId . assetNameHex
tokenSubject:    policyId   assetNameHex
```

Comment (wolverineks) [reply src/yoroi-wallets/cardano/api/utils.ts:64]: ultimately i want to deprecate `tokenIdentifier`, `tokenSubject`, and `tokenId`
and use `token.id = toAssetFingerprint(policyId, assetNameHex)`

**Pull request merged**